Ryzen model gen

poganesh opened the pull request:
## Describe your changes

## Checklist before requesting a review
- [ ] Add unit tests for this change.
- [ ] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [ ] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
- [ ] Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

## (Optional) Issue link

VishalX commented:
Adding some review comments directly here:

- Add AMD disclaimers to new/modified files.
- Use "vitisai" folder which already existing instead of adding new "ryzen"/"ryzenai" folder. Same for file naming.
- Why so many files being added? Intentional?

@poganesh

VishalX reviewed (changes requested):
Adding some review comments directly here:

- Add AMD disclaimers to new/modified files.
- Use "vitisai" folder which already existing instead of adding new "ryzen"/"ryzenai" folder. Same for file naming.
- Why so many files being added? Intentional?

poganesh commented:
@VishalX 

- AMD disclaimers are now added to new/modified files.
- Sticking to `vitisai` folder name instead of `ryzenai`
- Extra files were added for quark integration, however I have cleaned them up with minimal required files to support quark LLM awq.

VishalX commented on examples/llama2/README.md:
```diff
@@ -160,6 +160,56 @@ inputs = tokenizer("Hello, World", return_tensors="pt").to("cuda:0")
+##### **For Quark quantization**
+
+For LLMs - Follow the below commands to generate the optimized model for VitisAI EP.
+(Currently tested on Linux + ROCm and Linux + CUDA). 
```
This line is not very clear. Is this Linux + RoCM and Linux + CUDA both?

poganesh replied:
Updated the line to clarify that both Linux with ROCm and Linux with CUDA are supported. Also, added windows with cpu is supported. Please have a look now.

VishalX commented on examples/llama2/README.md:
````diff
@@ -160,6 +160,56 @@ inputs = tokenizer("Hello, World", return_tensors="pt").to("cuda:0")
+pip install -r requirements.txt
+```
+
+The model generate wheel can be downloaded from here (to be updated in PyPI)- `/proj/xsjhdstaff4/pooja/olive_wheels/model_generate-1.0.0-py3-none-manylinux2014_x86_64.whl` (Linux, python 3.10 wheel)
+
+```bash
+pip install "model_generate-1.0.0-py3-none-manylinux2014_x86_64.whl"
+```
````
Can this be installed using pip directly? Maybe with `--index-url` option?

VishalX replied:
```sh
pip install model_generate --index-url https://pypi.amd.com/simple/model-generate/
```

poganesh replied:
Added a pypi link for now with a suggested dummy --index-url. Also, added a note that it would be updated once the wheel is officially published in pypi.amd.com. Please review it.

VishalX commented on olive/passes/onnx/vitis_ai/examples/torch/language_modeling/llm_eval/.gitignore:
```diff
@@ -0,0 +1,4 @@
+/*.json
+!*/.json
+/*.txt
+!*/.txt
```
do we need this file? can you remove if not required?

poganesh replied:
Done. File has been deleted now.

VishalX commented on olive/passes/onnx/vitis_ai/examples/torch/language_modeling/llm_ptq/example_quark_torch_llm_ptq.rst:
```diff
@@ -0,0 +1,494 @@
+Language Model Post Training Quantization (PTQ) Using Quark
```
do we need this file to be added to olive?

poganesh replied:
It is not required, deleted in latest commit

ChaoLi-AMD commented on olive/passes/onnx/quark_quantization.py:
```diff
@@ -0,0 +1,114 @@
+#
```
QQ: since the LLM quantization is done by Quark Torch, wouldn’t it make more sense to put it under passes/pytorch?

thiagocrepaldi replied:
or maybe olive/passes/quark?

VishalX replied:
Agree. @poganesh can you do the suggested? 
Let's create quark folder : `olive/passes/quark` and move the files under that.

poganesh replied:
Done. Created quark_quantizer folder here `olive/passes/quark_quantizer` and added all quark files in it.

thiagocrepaldi commented on olive/passes/onnx/vitis_ai/examples/torch/language_modeling/llm_ptq/quantize_quark.py:
```diff
@@ -0,0 +1,461 @@
+    # TODO:
+    # The current method results in high CPU memory consumption due to multiple copies of the same model.
+    # We plan to address this in the future by implementing a more efficient way to dispatch the model to devices.
+    if args.use_tp:
```
should we default windows to cpu too?
```suggestion
    if args.use_tp or platform.system() == 'Windows':
```

must `import platform` beforehand

poganesh replied:
Done. Added the platform check and import platform at the top. Windows now defaults to CPU as suggested.

thiagocrepaldi commented on olive/passes/onnx/vitis_ai/examples/torch/language_modeling/llm_ptq/example_quark_torch_llm_ptq.rst:
```diff
@@ -0,0 +1,494 @@
+Language Model Post Training Quantization (PTQ) Using Quark
+===========================================================
```
why are we shipping `*.rst` files? they are quark documentation also available at https://quark.docs.amd.com

VishalX replied:
agree, no need to ship any .rst files

poganesh replied:
deleted the rst file - `llm_ptq/example_quark_torch_llm_ptq.rst`. No other `.rst` file exists.

thiagocrepaldi commented on examples/llama2/README.md:
````diff
@@ -160,6 +160,56 @@ inputs = tokenizer("Hello, World", return_tensors="pt").to("cuda:0")
 print(tokenizer.batch_decode(model.generate(**inputs, max_length=20), skip_special_tokens=True))
 ```
 
+### **Optimization and Quantization for AMD NPU**
````
do we need to ship *.md files?

poganesh replied:
`.md` files were already existing in Olive, we are only updating/adding instructions on how to run LLM on VitisAI EP using Olive. This is required.

thiagocrepaldi commented on examples/llama2/README.md:
````diff
@@ -160,6 +160,56 @@ inputs = tokenizer("Hello, World", return_tensors="pt").to("cuda:0")
 print(tokenizer.batch_decode(model.generate(**inputs, max_length=20), skip_special_tokens=True))
 ```
 
+### **Optimization and Quantization for AMD NPU**
````
is using `examples` name intended?

VishalX replied:
examples folder already exists in Olive, Pooja added/updated READMEs and recipes I think. @poganesh pls confirm

poganesh replied:
yes, examples were already existing in Olive. I have used the same folder structure to add our readme and config.

thiagocrepaldi commented on olive/passes/onnx/vitis_ai/examples/__init__.py:
```diff
@@ -0,0 +1,4 @@
+#
```
`examples` folder with `__init__.py` is odd. `examples` folder inside `passes` is also confusing. it seems we are using folder strucutre from quark examples into olive, which doesnt seem ideal

VishalX replied:
@poganesh do we use files from this `onnx/vitis_ai/examples` folder? If yes, can we rename it appropriately, let's not use `examples` for this.

poganesh replied:
Done. `vitis_ai/examples` has been removed. All quark files are now added in `Olive/olive/passes/quark_quantizer/torch`

VishalX commented on olive/passes/onnx/vitis_generate_model_llm.py:
```diff
@@ -0,0 +1,66 @@
+            input_model=str(input_model_path),
+            output_dir=str(output_dir),
+            packed_const=config.packed_const,
+            cpu_only=config.cpu_only,
```
Is this "cpu_only" flag added recently?

poganesh replied:
By default, it is set to False. If set to True CPU only OGA model is generated.

VishalX commented:
Close this?
